components/grouper/src/main/java/org/datacleaner/components/group/GrouperTransformer.java

```diff
@@ -0,0 +1,286 @@
+    AggregationType[] aggregationTypes;
+
+    @Configured(order = 4, value = PROPERTY_VALUE_SORTATION)
+    SortationType valueSortation = SortationType.NONE;
+
+    @Configured
+    String concatenationSeparator = ", ";
+
+    @Configured
+    boolean skipNullGroupKeys = true;
+
+    @Configured
+    boolean skipNullValues = true;
+
+    private final ConcurrentMap<Object, ConcurrentLinkedDeque<List<Object>>> _values = new ConcurrentHashMap<>();
+    private OutputRowCollector _rowCollector;
+
+    @Initialize
+    public void init() {
+        _values.clear();
+    }
+
+    @Override
+    public OutputDataStream[] getOutputDataStreams() {
+        final OutputDataStreamBuilder outputDataStreamBuilder = OutputDataStreams.pushDataStream("output");
+        outputDataStreamBuilder.withColumnLike(groupKey);
+        outputDataStreamBuilder.withColumn("row_count", ColumnType.INTEGER);
+        for (int i = 0; i < aggregatedValues.length; i++) {
+            final InputColumn<?> inputColumn = aggregatedValues[i];
+            final AggregationType aggregationType = aggregationTypes[i];
+            switch (aggregationType) {
+            case FIRST_VALUE:
+                outputDataStreamBuilder.withColumnLike(inputColumn);
+                break;
+            case CONCAT_VALUES:
+                outputDataStreamBuilder.withColumn(inputColumn.getName(), ColumnType.STRING);
+                break;
+            case CREATE_LIST:
+                outputDataStreamBuilder.withColumn(inputColumn.getName(), ColumnType.LIST);
+                break;
+            default:
+                throw new UnsupportedOperationException("Unsupported aggregation type: " + aggregationType);
+            }
+        }
+
+        final OutputDataStream stream = outputDataStreamBuilder.toOutputDataStream();
+        return new OutputDataStream[] { stream };
+    }
+
+    @Override
+    public void initializeOutputDataStream(OutputDataStream stream, Query q, OutputRowCollector collector) {
+        _rowCollector = collector;
+    }
+
+    @Override
+    protected void run(InputRow row) {
+        if (_rowCollector == null) {
+            // nothing to do
+            return;
+        }
+
+        Object key = row.getValue(groupKey);
+        if (key == null) {
+            if (skipNullGroupKeys) {
+                // skip it
+                return;
+            } else {
+                key = NULL_KEY;
+            }
+        }
+
+        final List<Object> values = row.getValues(aggregatedValues);
+        if (valueSortation == SortationType.RECORD_ORDER) {
+            // add row ID for the sortation
+            values.add(row.getId());
+        }
+
+        ConcurrentLinkedDeque<List<Object>> newCollectionOfValues = new ConcurrentLinkedDeque<>();
+        ConcurrentLinkedDeque<List<Object>> previousCollectionOfValues = _values
+                .putIfAbsent(key, newCollectionOfValues);
+        if (previousCollectionOfValues == null) {
```

Comment:
That would be a nice optimization for some cases (depends on combination of sortation and aggregation type). I guess we can cover some of those cases already, but we could also consider it a non-functional improvement/optimization.